networkx/algorithms/dag.py

```diff
@@ -170,10 +179,13 @@ def topological_sort(G):
 
     indegree_map = {v: d for v, d in G.in_degree() if d > 0}
     # These nodes have zero indegree and ready to be returned.
-    zero_indegree = [v for v, d in G.in_degree() if d == 0]
+    zero_indegree = deque(v for v, d in G.in_degree() if d == 0)
+
+    generation = 0
+    last_node_in_generation = zero_indegree[-1] if zero_indegree else None
 
     while zero_indegree:
-        node = zero_indegree.pop()
+        node = zero_indegree.popleft()
         if node not in G:
             raise RuntimeError("Graph changed during iteration")
         for _, child in G.edges(node):
```

Comment:
```suggestion
      generation += 1
      this_generation = zero_indegree
      zero_indegree = []
      for node in this_generation:
        if node not in G:
            raise RuntimeError("Graph changed during iteration")
        for child in G.neighbors(node):
```

Reply:
So I actually like this approach, but this then deviates from the referenced algorithm described in the docstring:
```
    .. [1] Manber, U. (1989).       *Introduction to Algorithms - A Creative Approach.* Addison-Wesley.
```
Screenshot from the book below.
![image](https://user-images.githubusercontent.com/13501084/118594035-71cf4400-b744-11eb-98f3-9de44e20d6ea.png)

Are we ok deviating from this?

Reply:
Does it deviate from the referenced algorithm?
I am thinking that this just changes the type of queue we are using (to a sequence of lists).
That is "put w in Queue" is changed to use a sequence of lists to provide the queue. Then "remove vertex v from Queue" removes the nodes from consideration one by one off the queue (doesn't remove them from the data structure until all nodes from that part of the queue are processed, but it does remove them from consideration).

As I was thinking about the bigger structure of these two functions last night, I notice that your new function recreates each of these lists. Would it be more effective to rearrange the computation to have the second (new) function generate the sets of nodes for each generation and then have the first function use that and then yield each of the nodes? Thoughts?

Reply:
The rearranged pattern you describe is actually the way I've done this in the past and what I would prefer too(the building the generations after the fact was solely to not disturb the original implementation, since I wasn't sure it would be appropriate for me to change a fundamental function that's been around for several years). If we're ok with updating the topological_sort, then I would prefer the way you described here: So would essentially look like this:

```
def topological_generations(G):
    indegree_map = {v: d for v, d in G.in_degree() if d > 0}
    zero_indegree = [v for v, d in G.in_degree() if d == 0]

    generation = -1

    while zero_indegree:
        generation += 1
        this_generation = zero_indegree
        zero_indegree = []
        for node in this_generation:
            if node not in G:
                raise RuntimeError("Graph changed during iteration")
            for child in G.neighbors(node):
                try:
                    indegree_map[child] -= 1
                except KeyError as e:
                    raise RuntimeError("Graph changed during iteration") from e
                if indegree_map[child] == 0:
                    zero_indegree.append(child)
                    del indegree_map[child]
        yield this_generation


def topological_sort(G):
    for generation in nx.topological_generations(G):
        for node in generation:
            yield node

```

Reply:
Yes -- That makes more sense to me than the first implementation. I'm glad your instincts are similar -- that makes it more likely to be a good way to design it. :}

we don't even need the two lines of `generation = -1` and `generation += 1`, because the user can get the generation number from the iterator with `enumerate(topological_generations)`. 

I think this is also significantly faster than the previous implementation (3-4x) -- and I think it leaves the API the same for the original function.
:}

Reply:
Cool, pushed the changes to the branch. All tests passing. Needed to change G.neighbors back to G.edges, since it was failing in cases of multiple edges between the same node.